### ISSUE TITLE
"you say MUST, but do you really mean it?"

### ISSUE THREAD

**Issue by martinthomson** [davidben/tls-key-share-prediction#5]
It took me a while to understand how the safe/unsafe distinction and the server behaviour were intended to interact.

The mechanism seems a bit convoluted, but it boils down to a client needing to include key shares for its most-preferred groups, unless it can be confident that the server will do the right thing when a more-preferred group lacks a share (which is what this draft attempts to mandate).

I would prefer not to include this mandate.  There are lots of reasons that this "MUST" might be ignored.

Mostly, it's just that an extra round trip is a giant pain. If you have a non-PQ share, but might prefer to use a hybrid-PQ KEM, you are faced with a choice: is your preference for hybrid-PQ greater than your preference for faster connection setup?  Do you need to factor in the current server load?

Ultimately, I would prefer to decouple the solution to this problem from the hinting stuff as much as possible.  I know that SVCB hints really amplify the risks, but I just don't see this going anywhere other than into RFC 6919 territory.

**Comment by davidben** [comment]
Yeah, I struggled a lot with how to phrase this in a remotely coherent way. I guess, for the client, the main question is when is it safe to actually exercise RFC 8446's allowance to omit your most preferred groups in key shares. Which boils down to whether you believe OpenSSL's existing behavior is:

1. An undesirable behavior that we account for because it's very common and, prior to key share prediction, there wasn't actually a problem so it'd be bad to introduce one
2. An undesirable behavior we can ignore because it's a security bug and the server operator's responsibility to patch
3. An explicit decision to treat all groups as equipreference, despite the documentation clearly describing otherwise

At the time I wrote this spec, I had interpreted it as (1) because (2) seemed a bit harsh, and would require clear consensus that this behavior is wrong. If we believe it's (1), I think we have to believe we need some kind of classification here.

I didn't give much thought to (3) because the documentation clearly said otherwise. But, on reflection, I'm warming up to that option:
* RFC 8446 already explicitly said "However, the values MAY be a non-contiguous subset of the "supported_groups" extension and MAY omit the most preferred groups". While it's kind of buried in there (I missed it at first!) and 8446bis's clarification is a much needed improvement, the spec was not *actually* ambiguous
* OpenSSL's position seems to be (3) right now, despite the documentation mismatch.

So perhaps the thing to do is I'll go write OpenSSL's documentation fix for them, and then yeah take interpretation (3). Maybe include some very loose text here to the effect of "clients, be aware that varying your key share set may allow attacker influence over the selected group... however, any influence will be within the confines of the server's selection policies. It is up to the server to not implement a stupid policy."
